sdk/python/feast/infra/offline_stores/contrib/trino_offline_store/connectors/upload.py

```diff
@@ -0,0 +1,189 @@
+"""
+Connectors can be found in the following doc https://trino.io/docs/current/connector.html
+
+For the hive connector, all file formats are here
+https://trino.io/docs/current/connector/hive.html#supported-file-types
+
+Example yaml config
+```yaml
+offline_store:
+    type: trino
+    host: localhost
+    port: 8080
+    catalog: hive
+    dataset: ci
+    connector:
+        type: hive
+        file_format: parquet
+```
+"""
+from datetime import datetime
+from typing import Any, Dict, Iterator, Optional, Set
+
+import numpy as np
+import pandas as pd
+import pyarrow
+from pytz import utc
+
+from feast.infra.offline_stores.contrib.trino_offline_store.trino_queries import Trino
+from feast.infra.offline_stores.contrib.trino_offline_store.trino_type_map import (
+    pa_to_trino_value_type,
+)
+
+CONNECTORS_DONT_SUPPORT_CREATE_TABLE: Set[str] = {
+    "druid",
+    "elasticsearch",
+    "googlesheets",
+    "jmx",
+    "kafka",
+    "kinesis",
+    "localfile",
+    "pinot",
+    "postgresql",
+    "prometheus",
+    "redis",
+    "thrift",
+    "tpcds",
+    "tpch",
+}
+CONNECTORS_WITHOUT_WITH_STATEMENTS: Set[str] = {
+    "bigquery",
+    "cassandra",
+    "memory",
+    "mongodb",
+    "mysql",
+    "oracle",
+    "redshift",
+    "memsql",
+}
+
+CREATE_SCHEMA_QUERY_TEMPLATE = """
+    CREATE TABLE IF NOT EXISTS {table} (
+        {schema}
+    )
+    {with_statement}
+"""
+
+INSERT_ROWS_QUERY_TEMPLATE = """
+    INSERT INTO {table} ({columns})
+    VALUES {values}
+"""
+
+
+def pyarrow_schema_from_dataframe(df: pd.DataFrame) -> Dict[str, Any]:
+    pyarrow_schema = pyarrow.Table.from_pandas(df).schema
+    trino_schema: Dict[str, Any] = {}
+    for field in pyarrow_schema:
+        try:
+            trino_type = pa_to_trino_value_type(str(field.type))
+        except KeyError:
+            raise ValueError(
+                f"Not supported type '{field.type}' in entity_df for '{field.name}'."
+            )
+        trino_schema[field.name] = trino_type
+    return trino_schema
+
+
+def trino_table_schema_from_dataframe(df: pd.DataFrame) -> str:
+    return ",".join(
+        [
+            f"{field_name} {field_type}"
+            for field_name, field_type in pyarrow_schema_from_dataframe(df=df).items()
+        ]
+    )
```

Comment:
small suggestion
```suggestion
def trino_table_schema_from_dataframe(df: pd.DataFrame) -> str:
    return ",".join(
        f"{field_name} {field_type}"
        for field_name, field_type in pyarrow_schema_from_dataframe(df=df).items()
    )
```

Reply:
This is done automatically by the formatter. Don't have control over it

Reply:
I removed the list concatenation, in favor of the generator syntax